keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1503 @@
+- (to be evaluated after alpha) allow to implement sidecar containers that will
+  guaranteed to be running longer than regular containers 
+
+### Non-Goals
+
+This proposal doesn't aim to:
+- support arbitrary dependencies graphs between containers
+- act as a security control to enforce that pod containers only run while the
+  sidecar is healthy. Restart of sidecar containers is a best effort
+- allow to enforce security boundaries to sidecar containers different that
+  other containers. For example, allow Istio to run privileged to configure ip
+  tables and disable other containers from doing so
+- (alpha) support containers termination ordering
+
+## Proposal
+
+The proposal is to introduce a `restartPolicy` field to Init containers and use
+it to indicate that an Init container is a sidecar container. Kubelet will start Init containers with
+`restartPolicy=Always` in the order with other Init containers, but instead of
+waiting for its completion, it will wait for the container startup completion.
+
+The condition for startup completion is open question. It is UNRESOLVED now
+and will be either:
+- pod is ready (readiness probe succeeded), or
+- startup probe succeeded or if no startup probe defined, `postStart` handler
+  completed, or
+- `postStart` handler completed.
+
+The decision on specific signal for container startup completion will
+be locked on before implementation will start. See the section
+["Pod startup completed condition"](#pod-startup-completed-condition)
+
+As part of the KEP, init containers and regular containers will be split into
+two different types. The field `restartPolicy` will only be introduced on init
+containers. The only supported value proposed in this KEP is `Always`. No other
+values will be defined as part of this KEP. Moreover, the field will be
+nullable so the default value will be "no value". 
+
+Other values for `restartPolicy` of containers will not be accepted and
+containers will follow the logic is currently implemented (documented
+[here](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#:~:text=if%20a%20pod's%20init%20container%20fails%2C%20the%20kubelet%20repeatedly%20restarts%20that%20init%20container%20until%20it%20succeeds.%20however%2C%20if%20the%20pod%20has%20a%20restartpolicy%20of%20never%2C%20and%20an%20init%20container%20fails%20during%20startup%20of%20that%20pod%2C%20kubernetes%20treats%20the%20overall%20pod%20as%20failed.)
+and more details can be found in the section ["Future use of `restartPolicy` field"]((#future-use-of--field))).
+
+Sidecar containers will not block Pod completion - if all regular containers
+complete, sidecar containers will be terminated.
+
+During the sidecar startup stage the restart behavior will be similar to Init
+containers. If Pod `restartPolicy` is `Never`, sidecar container failed during
+startup will NOT be restarted and the whole Pod will fail. If Pod
+`restartPolicy` is `Always` or `OnFailure`, it will be restarted.
```

Comment:
On the topic of per-container restartPolicy, in the last SIG-node meeting, I had wondered if we should bring the newly added per-container resizePolicy.restartPolicy under the restartPolicy proposed here. On further consideration, it is best not to conflate them. Per-container resizePolicy influences the outcome of ContainersToStart & ContainersToKill in `computePodActions` during resize and can operate orthogonal to per-container restartPolicy with the same validation constraints as defined today.

And for simplicity, we can perhaps build it such that if per-container restartPolicy is specified, it supercedes the pod-level restartPolicy value, JMHO.

Reply:
> As for adding a new field PodStatus.ResourcesAllocated, I don't favor it as currently proposed in this KEP. The ResourcesAllocated field is used to describe the resource levels at which each container in the pod has been admitted by a node (a.k.a achieved successful canAdmitPod(...)). This is not applicable to unassigned or Pending pods.

Once the pod is scheduled to the node, `PodStatus.ResourcesAllocated`, would be this value.  I can see the argument that its misusing the name "Allocated" prior to being scheduled, but I think the benefit of clearly showing the resources required for the pod to schedule outweighs that.

> IIRC, pending pods that are not scheduled due to lack of resources have their pod conditions updated with the underlying reasons. I'm also struggling to see how adding this field equips users with something actionable (programmatically) that can justify its presence in the pod api.

The condition is updated, however the message isn't helpful for determining how to solve the problem.  E.g. Below I know that I have insufficient CPU in this case, but how insufficient?  

```
  - lastProbeTime: null
    lastTransitionTime: "2023-01-31T15:23:24Z"
    message: '0/2 nodes are available: 1 Insufficient cpu, 1 node(s) had taint {CriticalAddonsOnly:
      true}, that the pod didn''t tolerate.'
    reason: Unschedulable
    status: "False"
    type: PodScheduled
```

If I wanted to determine how much more CPU the node would need so that this pod can run, I need to determine the resources required by the pod.  This requires the user to manually implement the multi-part calculation described in the KEP.

Reply:
@tzneal If possible, I want to reserve the name `PodStatus.ResourcesAllocated` for a possible future pod-level requests+limits KEP. On further thought, neither kubectl describe nor extending pod condition message are good options. We should respond with objects. `PodStatus.TotalResourcesRequested` better represents what you need as well as what metrics needs. Please see comment: https://github.com/kubernetes/enhancements/pull/3761#discussion_r1092772324 . WDYT?

Reply:
That seems fine to me, I chose `ResourcesAllocated` to avoid a proliferation of names even though its not precise, but if you suspect there's a need to use it for something else it makes sense to pick something better here.

Reply:
Besides the possibility of the pod level requests+limits KEP, we have defined ResourcesAllocated in the [In-place pod resize KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#api-changes) as the amount of resources node has allocated for the pod. It is odd to overload the _Allocated_ term for pending pods, (no resources allocated yet) and for pods that desire resize but node has not granted (Infeasible, Deferred, Proposed resize statuses).

Suggestion: If it is fine with you, could you please work with @SergeyKanzhelev to add this particular use case to the 'Goals of exposing the Pod.ResourceAllocated field' section. Giving the ability to cluster autoscaler, karpenter etc to collect a bunch of resource unavailable/unschedulable pods and easily sum up their TotalResourcesRequested to add nodes is a very strong justification to have this field in PodStatus. I only ask that <begin shameless plug> you use in-place resize to right-size over-provisioned pods first to help keep carbon footprint low and save some $ while at it before adding nodes :) <end shameless plug>

Reply:
added word to word, thank you!